File: packages/sdk/sdk-ruby/tests/api_key_mutations_test.rb
require "minitest/autorun"
require_relative "../lib/index"

class ApiKeyMutationsTest < Minitest::Test
  def test_api_key_mutation_helpers_return_payloads
    client = PhaseoSdk::Phaseo.new(
      api_key: "test",
      enable_deprecation_warnings: false
    )

    calls = []
    client.raw_client.define_singleton_method(:request) do |method:, path:, query: nil, headers: nil, body: nil|
      calls << [method, path, query, headers, body]
      case [method, path]
      when ["POST", "/keys"]
        { "data" => { "id" => "key_123", "name" => "Admin Key", "status" => "active" } }
      when ["PATCH", "/keys/key_123"]
        { "data" => { "id" => "key_123", "name" => "Renamed Key", "status" => "disabled" } }
      when ["DELETE", "/keys/key_123"]
        { "data" => { "id" => "key_123", "deleted" => true } }
      else
        raise "Unexpected request"
      end
    end

    created = client.create_api_key("name" => "Admin Key", "scopes" => ["gateway:read"])
    updated = client.update_api_key("key_123", "name" => "Renamed Key", "disabled" => true)
    deleted = client.delete_api_key("key_123")

    assert_equal "active", created["data"]["status"]
    assert_equal "disabled", updated["data"]["status"]
    assert_equal true, deleted["data"]["deleted"]
    assert_equal [
      ["POST", "/keys", nil, nil, { "name" => "Admin Key", "scopes" => ["gateway:read"] }],
      ["PATCH", "/keys/key_123", nil, nil, { "name" => "Renamed Key", "disabled" => true }],
      ["DELETE", "/keys/key_123", nil, nil, nil]
    ], calls
  end
end

File: packages/sdk/sdk-php/tests/api_key_mutations_test.php
<?php
declare(strict_types=1);

require_once __DIR__ . "/../src/index.php";

use Phaseo\Sdk\Phaseo;

function assert_true(bool $condition, string $message): void
{
    if (!$condition) {
        throw new RuntimeException($message);
    }
}

final class FakeApiKeyMutationsClient extends \Phaseo\Gen\Client
{
    public array $calls = [];

    public function __construct()
    {
        parent::__construct("https://example.test");
    }

    public function request(string $method, string $path, ?array $query = null, ?array $headers = null, $body = null)
    {
        $this->calls[] = [$method, $path, $query, $headers, $body];
        return match ([$method, $path]) {
            ["POST", "/keys"] => ["data" => ["id" => "key_123", "name" => "Admin Key", "status" => "active"]],
            ["PATCH", "/keys/key_123"] => ["data" => ["id" => "key_123", "name" => "Renamed Key", "status" => "disabled"]],
            ["DELETE", "/keys/key_123"] => ["data" => ["id" => "key_123", "deleted" => true]],
            default => throw new RuntimeException("Unexpected request"),
        };
    }
}

$client = new Phaseo(
    apiKey: "test",
    basePath: "https://api.phaseo.app/v1",
    enableDeprecationWarnings: false
);

$fake = new FakeApiKeyMutationsClient();
$ref = new ReflectionClass($client);
$prop = $ref->getProperty("client");
$prop->setAccessible(true);
$prop->setValue($client, $fake);

$created = $client->createApiKey(["name" => "Admin Key", "scopes" => ["gateway:read"]]);
$updated = $client->updateApiKey("key_123", ["name" => "Renamed Key", "disabled" => true]);
$deleted = $client->deleteApiKey("key_123");

assert_true(($created["data"]["status"] ?? null) === "active", "expected created key status");
assert_true(($updated["data"]["status"] ?? null) === "disabled", "expected updated key status");
assert_true(($deleted["data"]["deleted"] ?? null) === true, "expected deleted key payload");
assert_true($fake->calls === [
    ["POST", "/keys", null, null, ["name" => "Admin Key", "scopes" => ["gateway:read"]]],
    ["PATCH", "/keys/key_123", null, null, ["name" => "Renamed Key", "disabled" => true]],
    ["DELETE", "/keys/key_123", null, null, null],
], "expected request call sequence for api key mutations");

echo "php api key mutation tests ok" . PHP_EOL;
